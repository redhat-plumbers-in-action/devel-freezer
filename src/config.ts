import {
  ArrayMinSize,
  IsArray,
  IsString,
  MinLength,
  validate,
  ValidateNested,
} from 'class-validator';
import { Context } from 'probot';

import { events } from './events';
import { ValidationFeedback } from './validation-feedback';

import { TConfigObject, TFeedback, TPolicyItem } from './types.d';

export class Config {
  @IsArray()
  @ValidateNested({ each: true })
  @ArrayMinSize(1)
  private _policy: PolicyItem[];

  constructor(config: TConfigObject) {
    this._policy = Array.isArray(config?.policy)
      ? config.policy.map(item => new PolicyItem(item))
      : [];
  }

  get policy() {
    return this._policy;
  }

  static async getConfig(
    context: {
      [K in keyof typeof events]: Context<typeof events[K][number]>;
    }[keyof typeof events]
  ) {
    const retrievedConfig = await context.config<TConfigObject>(
      'development-freeze.yml'
    );

    if (Config.isConfigEmpty(retrievedConfig)) {
      throw new Error(
        `Missing configuration. Please setup 'devel-freezer' Action using 'development-freeze.yml' file.`
      );
    }

    const config = new this(retrievedConfig as TConfigObject);

    return config;
  }

  static isConfigEmpty(config: TConfigObject | null | unknown) {
    return config === null;
  }

  static async validate(instance: Config) {
    const validationResult = await validate(instance, {
      whitelist: true,
      forbidNonWhitelisted: true,
    });

    const results = validationResult.map(error => {
      return ValidationFeedback.composeFeedbackObject(error);
    });

    return results;
  }
}

class PolicyItem {
  @IsString({ each: true })
  @MinLength(1, { each: true })
  private _tags: string[];

  @ValidateNested()
  private _feedback: Feedback;

  constructor(item: TPolicyItem) {
    this._tags = item.tags;
    this._feedback = new Feedback(item.feedback);
  }

  get tags() {
    return this._tags;
  }

  get feedback() {
    return this._feedback;
  }
}

class Feedback {
  @IsString()
  @MinLength(1)
  private _frozenState: string;

  @IsString()
  @MinLength(1)
  private _unFreezeState: string;

  constructor(feedback: TFeedback) {
    this._frozenState = feedback['frozen-state'];
    this._unFreezeState = feedback['unfreeze-state'];
  }

  get frozenState() {
    return this._frozenState;
  }

  get unFreezeState() {
    return this._unFreezeState;
  }
}
