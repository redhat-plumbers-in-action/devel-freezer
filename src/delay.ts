import { debug, warning } from '@actions/core';
import { setTimeout } from 'timers/promises';

export async function delay(seconds: number) {
  warning(`Delaying the action for ${seconds} seconds.`);

  await setTimeout(seconds * 1000);

  debug('Delay is over.');
}
