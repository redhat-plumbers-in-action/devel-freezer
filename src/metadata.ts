import { Context } from 'probot';

import { events } from './events';

interface IMetadataObject {
  tag: string | undefined;
  commentID: string | undefined;
}

export class Metadata {
  private _tag: IMetadataObject['tag'];
  private _commentID: IMetadataObject['commentID'];

  constructor(metadata: IMetadataObject) {
    this._tag = metadata?.tag ?? undefined;
    this._commentID = metadata?.commentID ?? undefined;
  }

  get tag() {
    return this._tag;
  }

  set tag(value: IMetadataObject['tag']) {
    this._tag = value;
  }

  get commentID() {
    return this._commentID;
  }

  set commentID(value: IMetadataObject['commentID']) {
    if (this._commentID === undefined) {
      this._commentID = value;
    }
  }

  async setMetadata(
    context: {
      [K in keyof typeof events]: Context<typeof events[K][number]>;
    }[keyof typeof events]
  ) {
    if (this.commentID !== undefined) {
      await MetadataController.setMetadata(
        Metadata.metadataCommentID,
        this.commentID ?? '',
        context as unknown as Context
      );
    }

    // TODO: clear tag when un-freezed
    await MetadataController.setMetadata(
      Metadata.metadataFreezingTag,
      this.tag ?? '',
      context as unknown as Context
    );
  }

  static readonly metadataFreezingTag = 'freezing-tag';
  static readonly metadataCommentID = 'comment-id';

  static async getMetadata(
    context: {
      [K in keyof typeof events]: Context<typeof events[K][number]>;
    }[keyof typeof events]
  ) {
    return new Metadata({
      tag: await (MetadataController.getMetadata(
        Metadata.metadataFreezingTag.toString(),
        context as unknown as Context
      ) as Promise<IMetadataObject['tag']>),
      commentID: await (MetadataController.getMetadata(
        Metadata.metadataCommentID.toString(),
        context as unknown as Context
      ) as Promise<IMetadataObject['commentID']>),
    });
  }
}

/**
 * Based on probot-metadata - https://github.com/probot/metadata
 */
// eslint-disable-next-line @typescript-eslint/no-extraneous-class
export class MetadataController {
  static readonly regex = /\n\n<!-- devel-freezer = (.*) -->/;

  static async getMetadata(key: string, context: Context) {
    const body =
      (await context.octokit.issues.get(context.issue())).data.body || '';

    const match = body.match(MetadataController.regex);

    if (match) {
      const data = JSON.parse(match[1]);
      return key ? data && data[key] : data;
    }
  }

  static async setMetadata(
    key: string,
    value: string,
    context: Context,
    issueNumber?: number
  ) {
    let body =
      (
        await context.octokit.issues.get(
          context.issue(issueNumber ? { issue_number: issueNumber } : {})
        )
      ).data.body || '';

    let data = {};

    body = body.replace(MetadataController.regex, (_, json) => {
      data = JSON.parse(json);
      return '';
    });

    if (!data) data = {};

    if (typeof key === 'object') {
      Object.assign(data, key);
    } else {
      (data as { [key: string]: string })[key] = value;
    }

    return context.octokit.issues.update(
      context.issue({
        body: `${body}\n\n<!-- devel-freezer = ${JSON.stringify(data)} -->`,
        ...(issueNumber ? { issue_number: issueNumber } : {}),
      })
    );
  }
}
