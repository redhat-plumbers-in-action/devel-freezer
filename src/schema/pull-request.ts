import { z } from 'zod';

export const labelDataSchema = z.object({
  name: z.string(),
});

export type LabelData = z.infer<typeof labelDataSchema>;

export const milestoneDataSchema = z.object({
  html_url: z.string(),
  number: z.number(),
  title: z.string(),
  description: z.string(),
  state: z.string(),
});

export type MilestoneData = z.infer<typeof milestoneDataSchema>;

export const pullRequestDataSchema = z.object({
  labels: z.array(labelDataSchema),
  milestone: milestoneDataSchema.nullable(),
});

export type PullRequestData = z.infer<typeof pullRequestDataSchema>;
