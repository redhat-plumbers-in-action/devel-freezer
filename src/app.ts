import { debug, error, warning } from '@actions/core';
import { Context, Probot } from 'probot';
import { validateOrReject } from 'class-validator';

import { events } from './events';
import { freezePullRequest } from './freezer';
import { getLatestTag, isTagFreezed } from './release';
import { Config } from './config';

const app = (probot: Probot) => {
  probot.on(
    events.pull_request,
    async (context: Context<typeof events.pull_request[number]>) => {
      const config = await Config.getConfig(context);

      await validateOrReject(config);

      if (!config) {
        error(
          `Missing configuration. Please setup 'devel-freezer' action using 'development-freeze.yml' file.`
        );
        return;
      }

      const latestTag = await getLatestTag(context);

      if (!latestTag) {
        warning(`Repository doesn't have any tags or releases published.`);
        return;
      }

      debug(`Latest tag is: '${latestTag}'`);

      for (const item of config.policy) {
        if (!isTagFreezed(latestTag, item.tags)) {
          continue;
        }

        await freezePullRequest(item.feedback.freezedState, latestTag, context);
        return;
      }

      // TODO: If PR is changed and contains metadata from devel-freezer, update comment
      // ? It requires knowledge about old freezing tag ...

      // get metadata if any
      // if metadata check tag and get unfreezing comment (for loop policy)
      // update metadata and comment

      debug(
        `The latest tag doesn't match the requirements for a development freeze.`
      );
    }
  );
};

export default app;
