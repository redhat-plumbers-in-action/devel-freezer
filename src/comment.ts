import { Context } from 'probot';
import { debug } from 'console';

import { events } from './events';
import { getMetadata, setMetadata } from './metadata';
import { warning } from '@actions/core';

const develFreezerCommentID = 'devel-freezer-comment-id';
const develFreezerFreezingTag = 'devel-freezer-freezing-tag';

export async function publishComment(
  content: string,
  tag: string,
  context: {
    [K in keyof typeof events]: Context<typeof events[K][number]>;
  }[keyof typeof events]
) {
  const commentID = await getCommentID(context);

  if (commentID) {
    updateComment(commentID, content, context);
    return;
  }

  const commentPayload = (await createComment(content, context))?.data;

  if (!commentPayload) {
    warning(`Failed to create comment.`);
    return;
  }

  // Set metadata
  await setCommentID(commentPayload.id, context);
  // TODO: Probably dont work when updating comment!!!
  await setFeezingTag(tag, context);
}

async function getCommentID(
  context: {
    [K in keyof typeof events]: Context<typeof events[K][number]>;
  }[keyof typeof events]
): Promise<number | undefined> {
  return getMetadata(develFreezerCommentID, context as unknown as Context);
}

async function setCommentID(
  id: number,
  context: {
    [K in keyof typeof events]: Context<typeof events[K][number]>;
  }[keyof typeof events]
) {
  debug(`id: '${id}'`);

  await setMetadata(
    develFreezerCommentID,
    id.toString(),
    context as unknown as Context
  );
}

async function setFeezingTag(
  tag: string,
  context: {
    [K in keyof typeof events]: Context<typeof events[K][number]>;
  }[keyof typeof events]
) {
  debug(`tag: '${tag}'`);

  await setMetadata(
    develFreezerFreezingTag,
    tag.toString(),
    context as unknown as Context
  );
}

function createComment(
  body: string,
  context: {
    [K in keyof typeof events]: Context<typeof events[K][number]>;
  }[keyof typeof events]
) {
  if (!body || body === '') return;

  return context.octokit.issues.createComment(
    context.issue({
      body,
    })
  );
}

async function updateComment(
  commentID: number,
  body: string,
  context: {
    [K in keyof typeof events]: Context<typeof events[K][number]>;
  }[keyof typeof events]
) {
  return context.octokit.issues.updateComment(
    context.issue({
      comment_id: commentID,
      body,
    })
  );
}
