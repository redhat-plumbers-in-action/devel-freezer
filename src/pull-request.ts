import { warning } from '@actions/core';
import { Context } from 'probot';

import { events } from './events';
import { Metadata } from './metadata';

export class PullRequest {
  private _metadata: Metadata;

  constructor(metadata: Metadata) {
    this._metadata = metadata;
  }

  get metadata() {
    return this._metadata;
  }

  isFreezed() {
    return !!this.metadata.commentID && !!this.metadata.tag;
  }

  isTagPolicyCompliant(tagPolicy: string[], tag?: string) {
    const freezingTag = tag ?? this.metadata.tag;
    if (freezingTag === undefined) false;

    return tagPolicy.some(regex =>
      new RegExp(regex).test(freezingTag as string)
    );
  }

  async freeze(
    content: string,
    freezingTag: string,
    context: {
      [K in keyof typeof events]: Context<typeof events[K][number]>;
    }[keyof typeof events]
  ) {
    const id = await this.publishComment(content, context);

    this.metadata.commentID = id === undefined ? id : id.toString();
    this.metadata.tag = freezingTag;
    await this.metadata.setMetadata(context);
  }

  async unfreeze(
    content: string,
    context: {
      [K in keyof typeof events]: Context<typeof events[K][number]>;
    }[keyof typeof events]
  ) {
    const id = await this.publishComment(content, context);

    this.metadata.commentID = id === undefined ? id : id.toString();
    await this.metadata.setMetadata(context);
  }

  private async publishComment(
    content: string,
    context: {
      [K in keyof typeof events]: Context<typeof events[K][number]>;
    }[keyof typeof events]
  ) {
    if (this.metadata.commentID) {
      this.updateComment(content, context);
      return;
    }

    const commentPayload = (await this.createComment(content, context))?.data;

    if (!commentPayload) {
      warning(`Failed to create comment.`);
      return;
    }

    return commentPayload.id;
  }

  private createComment(
    body: string,
    context: {
      [K in keyof typeof events]: Context<typeof events[K][number]>;
    }[keyof typeof events]
  ) {
    if (!body || body === '') return;

    return context.octokit.issues.createComment(
      context.issue({
        body,
      })
    );
  }

  private async updateComment(
    body: string,
    context: {
      [K in keyof typeof events]: Context<typeof events[K][number]>;
    }[keyof typeof events]
  ) {
    if (!this.metadata.commentID) return;

    return context.octokit.issues.updateComment(
      context.issue({
        comment_id: +this.metadata.commentID,
        body,
      })
    );
  }

  static async getPullRequest(
    context: {
      [K in keyof typeof events]: Context<typeof events[K][number]>;
    }[keyof typeof events]
  ) {
    return new PullRequest(await Metadata.getMetadata(context));
  }
}
