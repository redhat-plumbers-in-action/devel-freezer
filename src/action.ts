import { debug, warning, getInput } from '@actions/core';

import { Config } from './config';
import { CustomOctokit } from './octokit';
import { PullRequest } from './pull-request';
import { Tag } from './tag';
import { delay } from './delay';

import { inputDelaySchema } from './schema/inputs';

async function action(octokit: CustomOctokit): Promise<void> {
  const delayParsed = inputDelaySchema.safeParse(getInput('delay'));
  const delaySeconds = delayParsed.success ? delayParsed.data : 0;

  if (delaySeconds > 0) {
    await delay(delaySeconds);
  }

  const config = await Config.getConfig(octokit);

  const tag = new Tag(await Tag.getLatestTag());

  if (!tag.latest) {
    warning(`Repository doesn't have any tags or releases published.`);
    return;
  }

  debug(`Latest tag is: '${tag.latest}'`);

  const pullRequest = await PullRequest.getPullRequest(
    +getInput('pr-number'),
    octokit
  );

  for (const policyItem of config.policy) {
    if (!tag.isFreezed(policyItem.tags)) {
      continue;
    }

    await pullRequest.freeze(policyItem.feedback['frozen-state'], tag.latest);
    return;
  }

  if (!pullRequest.isFreezed()) {
    return;
  }

  for (const policyItem of config.policy) {
    if (!pullRequest.isTagPolicyCompliant(policyItem.tags)) {
      continue;
    }

    await pullRequest.unfreeze(policyItem.feedback['unfreeze-state']);
    return;
  }

  debug(
    `The latest tag doesn't match the requirements for a development freeze.`
  );
}

export default action;
