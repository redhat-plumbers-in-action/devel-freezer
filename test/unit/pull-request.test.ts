import { describe, it, expect, beforeEach, test } from 'vitest';

import {
  pullRequestContextFixture,
  IPullRequestTestContext,
} from './fixtures/pull-request.fixture';

import { Metadata } from '../../src/metadata';
import { PullRequest } from '../../src/pull-request';

describe('Pull Request Object', () => {
  beforeEach<IPullRequestTestContext>(context => {
    context.pullRequests = pullRequestContextFixture.pullRequests;
    context.invalid = pullRequestContextFixture.invalid;
  });

  it<IPullRequestTestContext>('can be instantiated', context =>
    context.pullRequests.map(pullRequestItem =>
      expect(pullRequestItem).toBeDefined()
    ));

  test<IPullRequestTestContext>('get metadata()', context =>
    context.pullRequests.map(pullRequestItem =>
      expect(pullRequestItem.metadata).toMatchSnapshot()
    ));

  test<IPullRequestTestContext>('isFreezed()', context =>
    context.pullRequests.map(pullRequestItem =>
      expect(pullRequestItem.isFreezed()).toMatchSnapshot()
    ));

  test<IPullRequestTestContext>('isTagPolicyCompliant()', context => {
    context.pullRequests.map(pullRequestItem => {
      expect(
        pullRequestItem.isTagPolicyCompliant(['alpha', 'beta', 'aaa'])
      ).toMatchSnapshot();
      expect(
        pullRequestItem.isTagPolicyCompliant(['^\\S*-rc\\d$'])
      ).toMatchSnapshot();
    });

    const test_pr = new PullRequest(
      new Metadata({ tag: undefined, commentID: undefined })
    );

    // policy with list of tags
    expect(
      test_pr.isTagPolicyCompliant(['alpha', 'beta', 'aaa'], 'beta')
    ).toEqual(true);
    expect(
      test_pr.isTagPolicyCompliant(['alpha', 'beta', 'aaa'], 'bbb')
    ).toEqual(false);

    // policy with regular expressions
    expect(test_pr.isTagPolicyCompliant(['^\\S*-rc\\d$'], 'v250-rc1')).toEqual(
      true
    );
    expect(test_pr.isTagPolicyCompliant(['^\\S*-rc\\d$'], 'v250')).toEqual(
      false
    );
    expect(test_pr.isTagPolicyCompliant(['^\\S*-rc\\d$'], undefined)).toEqual(
      false
    );
  });
});
