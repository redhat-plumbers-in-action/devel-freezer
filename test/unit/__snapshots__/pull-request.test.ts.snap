// Vitest Snapshot v1

exports[`Pull Request Object > get metadata() 1`] = `
Metadata {
  "_commentID": undefined,
  "_tag": undefined,
}
`;

exports[`Pull Request Object > get metadata() 2`] = `
Metadata {
  "_commentID": undefined,
  "_tag": "v1",
}
`;

exports[`Pull Request Object > get metadata() 3`] = `
Metadata {
  "_commentID": "123456789",
  "_tag": undefined,
}
`;

exports[`Pull Request Object > get metadata() 4`] = `
Metadata {
  "_commentID": "123456789",
  "_tag": "v1",
}
`;

exports[`Pull Request Object > get metadata() 5`] = `
Metadata {
  "_commentID": undefined,
  "_tag": "alpha",
}
`;

exports[`Pull Request Object > get metadata() 6`] = `
Metadata {
  "_commentID": undefined,
  "_tag": "beta",
}
`;

exports[`Pull Request Object > get metadata() 7`] = `
Metadata {
  "_commentID": undefined,
  "_tag": "v250-rc1",
}
`;

exports[`Pull Request Object > isFreezed() 1`] = `false`;

exports[`Pull Request Object > isFreezed() 2`] = `false`;

exports[`Pull Request Object > isFreezed() 3`] = `false`;

exports[`Pull Request Object > isFreezed() 4`] = `true`;

exports[`Pull Request Object > isFreezed() 5`] = `false`;

exports[`Pull Request Object > isFreezed() 6`] = `false`;

exports[`Pull Request Object > isFreezed() 7`] = `false`;

exports[`Pull Request Object > isTagPolicyCompliant() 1`] = `false`;

exports[`Pull Request Object > isTagPolicyCompliant() 2`] = `false`;

exports[`Pull Request Object > isTagPolicyCompliant() 3`] = `false`;

exports[`Pull Request Object > isTagPolicyCompliant() 4`] = `false`;

exports[`Pull Request Object > isTagPolicyCompliant() 5`] = `false`;

exports[`Pull Request Object > isTagPolicyCompliant() 6`] = `false`;

exports[`Pull Request Object > isTagPolicyCompliant() 7`] = `false`;

exports[`Pull Request Object > isTagPolicyCompliant() 8`] = `false`;

exports[`Pull Request Object > isTagPolicyCompliant() 9`] = `true`;

exports[`Pull Request Object > isTagPolicyCompliant() 10`] = `false`;

exports[`Pull Request Object > isTagPolicyCompliant() 11`] = `true`;

exports[`Pull Request Object > isTagPolicyCompliant() 12`] = `false`;

exports[`Pull Request Object > isTagPolicyCompliant() 13`] = `false`;

exports[`Pull Request Object > isTagPolicyCompliant() 14`] = `true`;
